### ISSUE TITLE
Specific www-authenticate header causes mechanize to be stuck in a loop

### ISSUE THREAD

**Issue by gchan** [sparklemotion/mechanize#441]
It appears that mechanize gets stuck in an loop within the `Mechanize::HTTP::WWWAuthenticateParser#parse` method when it encounters a `www-authenticate` value of `Bearer realm='OAuth API', error='invalid_token'`

```
I, [2015-09-12T16:34:20.703320 #6267]  INFO -- : Net::HTTP::Get: [REDACTED]
D, [2015-09-12T16:34:20.703470 #6267] DEBUG -- : request-header: accept-encoding => gzip,deflate,identity
D, [2015-09-12T16:34:20.703575 #6267] DEBUG -- : request-header: accept => */*
D, [2015-09-12T16:34:20.703679 #6267] DEBUG -- : request-header: user-agent => Mozilla/5.0 (Windows NT 6.3; WOW64; rv:26.0) Gecko/20100101 Firefox/26.0
D, [2015-09-12T16:34:20.703779 #6267] DEBUG -- : request-header: accept-charset => ISO-8859-1,utf-8;q=0.7,*;q=0.7
D, [2015-09-12T16:34:20.703889 #6267] DEBUG -- : request-header: accept-language => en-us,en;q=0.5
D, [2015-09-12T16:34:20.703994 #6267] DEBUG -- : request-header: cookie => [REDACTED]=ak9JMWtlRmZSZ01QNW5SSld1cnpSY1VQTngrQkNiaFBLTUZ0cEloYkxrenBuRzY3Ui96S0lJRFRhcnNKV0x3c3VxazV2MHVwcEZTUEdxWVd0TGN6eE9jWDJ5WkVXQlhxTXJOb0xwYlZJaW89LS04N3U5K1NHajYrVmliU0JGSWEzMmhBPT0%3D--50e71170622c769a64c1ea370df5b7a178545f44
D, [2015-09-12T16:34:20.704096 #6267] DEBUG -- : request-header: host => [REDACTED].com
I, [2015-09-12T16:34:22.549526 #6267]  INFO -- : status: Net::HTTPUnauthorized 1.1 401 Unauthorized
D, [2015-09-12T16:34:22.549688 #6267] DEBUG -- : response-header: server => Cowboy
D, [2015-09-12T16:34:22.549795 #6267] DEBUG -- : response-header: connection => keep-alive
D, [2015-09-12T16:34:22.549899 #6267] DEBUG -- : response-header: date => Sat, 12 Sep 2015 04:34:22 GMT
D, [2015-09-12T16:34:22.550001 #6267] DEBUG -- : response-header: status => 401 Unauthorized
D, [2015-09-12T16:34:22.550105 #6267] DEBUG -- : response-header: strict-transport-security => max-age=31536000
D, [2015-09-12T16:34:22.550208 #6267] DEBUG -- : response-header: content-type => application/json
D, [2015-09-12T16:34:22.550310 #6267] DEBUG -- : response-header: x-accepted-oauth-scopes => ["investor"]
D, [2015-09-12T16:34:22.550420 #6267] DEBUG -- : response-header: www-authenticate => Bearer realm='OAuth API', error='invalid_token'
D, [2015-09-12T16:34:22.550522 #6267] DEBUG -- : response-header: content-length => 25
D, [2015-09-12T16:34:22.550623 #6267] DEBUG -- : response-header: cache-control => no-cache
D, [2015-09-12T16:34:22.550724 #6267] DEBUG -- : response-header: set-cookie => [REDACTED]=a2dQcm15WENZZHQ5NjZ0Mml5TzNnU20yakJMZ3FDeTAxcGJDN1FJNWM4YWJwZnNGNWdEck53c0VxM25pbWsydEJQWk5aRy91TGgvN1huWTlDN3VjSjByWGhIenRCN2laZkN2NUhYNVdxcGc9LS1EalpBUWVpR3h6MU5mdFdrbzdVL3BBPT0%3D--f32803fe734b40232d4be36cee453cae3b431496; domain=.[REDACTED].com; path=/; HttpOnly; secure
D, [2015-09-12T16:34:22.550826 #6267] DEBUG -- : response-header: x-request-id => 6f1d9ffa-3e03-4ffc-98c2-5001837c89de
D, [2015-09-12T16:34:22.550926 #6267] DEBUG -- : response-header: x-runtime => 0.013293
D, [2015-09-12T16:34:22.551032 #6267] DEBUG -- : response-header: via => 1.1 vegur
D, [2015-09-12T16:34:22.551304 #6267] DEBUG -- : Read 25 bytes (25 total)
D, [2015-09-12T16:34:22.553136 #6267] DEBUG -- : saved cookie: [REDACTED]=a2dQcm15WENZZHQ5NjZ0Mml5TzNnU20yakJMZ3FDeTAxcGJDN1FJNWM4YWJwZnNGNWdEck53c0VxM25pbWsydEJQWk5aRy91TGgvN1huWTlDN3VjSjByWGhIenRCN2laZkN2NUhYNVdxcGc9LS1EalpBUWVpR3h6MU5mdFdrbzdVL3BBPT0%3D--f32803fe734b40232d4be36cee453cae3b431496
^CInterrupt:
from /Users/gchan/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/bundler/gems/mechanize-4a011394d57d/lib/mechanize/http/www_authenticate_parser.rb
```

**Comment by gchan** [comment]
After some debugging, it turns out it the www-authenticate value `Bearer realm='OAuth API', error='invalid_token'` does not meet the RFC2617 and RFC2616 standard of a "quoted-string". 

`'OAuth API'` is surrounded with single quotes rather than double quote which is causing it to fail on this [line] (https://github.com/sparklemotion/mechanize/blob/master/lib/mechanize/http/www_authenticate_parser.rb#L133). The `#quoted_string` method probably does not support single quotes as well.

I'm not sure what we should do here? Should we support single quoted strings? Should we update the `parse` code so that I can't get stuck in a loop?

**Comment by zenspider** [comment]
This is way stale so I'm not sure if you're on board anymore... but my read of the current code is that if it doesn't get an auth_scheme (a token), it circles back around and grabs another. Am I misunderstanding something?

**Comment by gchan** [comment]
This still appears to be an issue (the test I've written in #442 still causes an infinite loop). If it doesn't get an auth_scheme the `StringScanner` doesn't proceed so circles back to try again and will fail again.

I've rebased my PR at #442 for your consideration.

**Comment by gchan** [comment]
Resolved by #442
